### ISSUE TITLE
KOKKOS_INLINE_FUNCTION on Serial execution space

### ISSUE THREAD

**Issue by kyungjoo-kim** [kokkos/kokkos#223]
When a user put an Intrepid function (or any library APIs which already contain parallel_for inside) in his parallel_for, serial execution space can be used that is allowed as it is simply a for-loop. In order to run the for-loop on device space, it needs KOKKOS_INLINE_FUNCTION decoration.

**Comment by crtrott** [comment]
Hi, what is the actual issue? I.e. is something not working, or do you need a new feature?

**Comment by hcedwar** [comment]
Ability to execute:
`parallel_for( RangePolicy<Serial>(0,N) , ... );`
from within an executing kernel, even on the device.
Should be as simple as adding KOKKOS_INLINE_FUNCTION to the Serial back-end specializations.

**Comment by kyungjoo-kim** [comment]
The use senario is that

template<ExecSpace>
void MyIntrepidFunction() {
   parallel_for(RangePolicy<ExecSpace>(0, N), MyFunctor);
}

And a user may want to call this in two different ways

main() { 
  MyIntrepidFunction<ThreadsSpace>();  // internal parallelism do
  parallel_for(RangePolicy<ThreadsSpace>(0, M), KOKKOS_LAMBDA() {
         MyIntrepidFunction<SerialSpace>();
  }
}

Then, a single code can be used in the outer parallel for if a user wants high level parallelization.

**Comment by crtrott** [comment]
Ok I see. Are you willing to template MyIntrepidFunction on the executiion space. I.e. make it:

```C++
MyIntrepidFunction<DefaultExecutionSpace>()
parallel_for(M, KOKKOS_LAMDBA () {
  MyIntrepidFunction<Serial>();
}
```

**Comment by kyungjoo-kim** [comment]
Yes of course. In the previous comment, I did put \<\> and github make them disappear (now I know that it needs escape character). Now I find out the buttom "insert code".

**Comment by kyungjoo-kim** [comment]
We might need to make other parallel policy object to be compiled on device. In the following pseudo code, range policy object cannot be compiled on device in the current version of kokkos. 

```
template<typename ExecSpace>
KOKKOS_INLINE_FUNCTION
void function(input, output) {
  RangePolicy<ExecSpace> policy(begin, end); // problem: no device version of range policy
  parallel_for(policy, functor); 
}

template<typename ExecSpace>
void doSomething() {
  RangePolicy<ExecSpace> policy(global_begin, global_end);
  parallel_for(policy, KOKKOS_LAMBDA() {
      input = subview(global_input);
      output = subview(global_output);
      function(input, output);
  }
}

int main() {
  function<DeviceSpace>(input, output);
  doSomething<DeviceSpace>(input, output);
}
```

**Comment by dholladay00** [comment]
What's the status of this?

I have a similar use case. I have 4 levels of parallelism and I'd like to easily switch which ones are parallelized:

```
parallel for (large number)
  serial for (large number)
    parallel reduce (small number)
      parallel reduce (large number)
```
to 

```
parallel for (large number)
  parallel for (large number)
    serial reduce (small number)
      parallel reduce (large number)
```

If I could write all of them with parallel patterns, rather than an explicit for loop, I could just change the policies around to test out different strategies. 

Currently fails on cuda backend when I try to make a `Kokkos::RangePolicy<Serial> p(0, N);` saying that `calling a __host__ function from a __host__ __device__ function is not allowed.`

**Comment by PhilMiller** [comment]
This actually ties in to a stated 'theme' for the coming FY, of self-similar dispatch APIs.

I'd suggest that rather than using `Kokkos::Serial` for nestable calls, we come up with a different name. My reasoning is that `Serial` is putatively a host-based execution space, and hence `HostSpace` is accessible to it. If it's something run on a GPU, that should not be the case.

(I happened to look at this, as the oldest open issue in the repository)

**Comment by ibaned** [comment]
@PhilMiller I agree with not using `Serial` for this. In my P3A library, I defined three execution policies (ISO C++17 parlance):

1. `execution::seq`: this is the "truly serial" policy, can be nested arbitrarily, is known to be defined as a non-GPU-executable for loop. Should be pretty much the same as using `std::execution::seq` from the standard library.
2. `execution::par`: this equivalent to `Kokkos::DefaultExecutionSpace`. Cannot be nested.
3. `execution::hot`: the name may be terrible, but this is defined as a `__host__ __device__` "serial" for loop, essentially a `Kokko::Serial`-like thing that can be called inside an `execution::par` loop, and can continue to be nested within itself.

I know that Kokkos has team parallelism too, but I would recommend adding the first and last as new execution spaces in Kokkos. Both arbitrarily nestable, one on the host outside of parallel dispatch and one inside a parallel dispatch thread.

**Comment by crtrott** [comment]
Couple things: we will not introduce a "sequential" execution policy as defined by std::execution:seq, semantically we will always require that iterations are independent and race free.

Yes we need a new thing for serial, we also need to deal with what it means to dispatch to the same "serial" instance.

**Comment by masterleinad** [comment]
> Yes we need a new thing for serial, we also need to deal with what it means to dispatch to the same "serial" instance.

Related to https://github.com/kokkos/kokkos-core-wiki/pull/388, https://github.com/kokkos/kokkos/pull/6051, and https://github.com/kokkos/kokkos/pull/6151.

**Comment by cwpearson** [comment]
#8367 lets you define a `parallel_for` that can be dispatched globally or nested withing a `TeamPolicy` (not a `RangePolicy`):

```c++
template <class Exec, class X, class Y>
KOKKOS_INLINE_FUNCTION void MyIntrepidFunction(const Exec& exec, X& x, Y& y) {
  auto policy = Kokkos::RangePolicy(exec, 0, x.extent(0));
  Kokkos::parallel_for(policy, KOKKOS_LAMBDA(const int& i) { x(i) += y(i); });
}
```
